osquery/bro/BrokerManager.cpp

```diff
@@ -0,0 +1,312 @@
+    return Status(1, "Broker Endpoint not set");
+  }
+
+  this->ep->peer(ip, port);
+  auto cs = this->ep->outgoing_connection_status().need_pop().front();
+  if (cs.status != broker::outgoing_connection_status::tag::established) {
+    LOG(ERROR) << "Failed to connect to bro endpoint";
+    return Status(1, "Failed to connect");
+  }
+
+  // Announce this endpoint to be a bro-osquery extension
+  // Collect Groups
+  broker::vector group_list;
+  for (std::string g : this->getGroups()) {
+    group_list.push_back(g);
+  }
+  // Collect IPs
+  broker::vector addr_list;
+  SQL sql("SELECT address from interface_addresses");
+  if (sql.ok()) {
+    for (const auto& row : sql.rows()) {
+      std::string if_mac = row.at("address");
+      addr_list.push_back(broker::address::from_string(if_mac).get());
+    }
+  } else {
+    return Status(1, "Failed to retrieve interface addresses");
+  }
+
+  // Create Message
+  broker::message announceMsg = broker::message{
+      this->EVENT_HOST_NEW, this->getNodeID(), group_list, addr_list};
+  this->sendEvent(TOPIC_ANNOUNCE, announceMsg);
+
+  return Status(0, "OK");
+}
+
+/////////////////////////////////////////////////////////
+//////////////// Broker Send Methods/////////////////////
+/////////////////////////////////////////////////////////
+
+Status BrokerManager::logQueryLogItemToBro(const QueryLogItem& qli) {
+  // Attributes from QueryLogItem
+  std::string queryID = qli.name; // The QueryID
+  //  std::string identifier = qli.identifier; // The HostID
+  //  size_t time = qli.time;
+  //  std::string calendar_time = qli.calendar_time;
+  //  std::map<std::string, std::string> decorations = qli.decorations;me;
+
+  // Is this schedule or one-time? Get Query and Type
+  std::string query;
+  std::string qType;
+  auto status_find = this->qm->findQueryAndType(queryID, qType, query);
+  if (!status_find.ok()) {
+    return status_find;
+  }
+
+  // Rows to be reported
+  std::vector<std::tuple<osquery::Row, std::string>> rows;
+  for (const auto& row : qli.results.added) {
+    rows.emplace_back(row, "ADD");
+  }
+  for (const auto& row : qli.results.removed) {
+    rows.emplace_back(row, "REMOVE");
+  }
+  for (const auto& row : qli.snapshot_results) {
+    rows.emplace_back(row, "SNAPSHOT");
+  }
+
+  // Get Info about SQL Query and Types
+  TableColumns columns;
+  Status status = getQueryColumns(query, columns);
+  std::map<std::string, ColumnType> columnTypes;
+  for (std::tuple<std::string, ColumnType, ColumnOptions> t : columns) {
+    std::string columnName = std::get<0>(t);
+    ColumnType columnType = std::get<1>(t);
+    //      ColumnOptions columnOptions = std::get<2>(t);
+    columnTypes[columnName] = columnType;
+    //    LOG(INFO) << "Column named '" << columnName << "' is of type '" <<
+    //    kColumnTypeNames.at(columnType) << "'";
+  }
+
+  // Common message fields
+  std::string uid = this->getNodeID();
+  std::string topic = this->qm->getEventTopic(queryID);
```

Comment:
As these are common message fields, they are equal for all messages that will be created in the following `for`.  I think it is better to set them once and use these local variable later for every message that is created